clients/pkg/promtail/targets/file/filetarget_test.go

```diff
@@ -336,6 +336,90 @@ func TestFileTarget_StopsTailersCleanly_Parallel(t *testing.T) {
 	ps.Stop()
 }
 
+// Make sure that Stop() doesn't hang if FileTarget is waiting on a channel send.
+func TestFileTarget_StopAbruptly(t *testing.T) {
+	w := log.NewSyncWriter(os.Stderr)
+	logger := log.NewLogfmtLogger(w)
+
+	dirName := newTestLogDirectories(t)
+	positionsFileName := filepath.Join(dirName, "positions.yml")
+	logDir1 := filepath.Join(dirName, "log1")
+	logDir2 := filepath.Join(dirName, "log2")
+	logDir3 := filepath.Join(dirName, "log3")
+
+	logfile1 := filepath.Join(logDir1, "test1.log")
+	logfile2 := filepath.Join(logDir2, "test1.log")
+	logfile3 := filepath.Join(logDir3, "test1.log")
+
+	ps, err := positions.New(logger, positions.Config{
+		SyncPeriod:    10 * time.Millisecond,
+		PositionsFile: positionsFileName,
+	})
+	require.NoError(t, err)
+
+	client := fake.New(func() {})
+	defer client.Stop()
+
+	// fakeHandler has to be a buffered channel so that we can call the len() function on it.
+	// We need to call len() to check if the channel is full.
+	fakeHandler := make(chan fileTargetEvent, 1)
+	pathToWatch := filepath.Join(dirName, "**", "*.log")
+	registry := prometheus.NewRegistry()
+	target, err := NewFileTarget(NewMetrics(registry), logger, client, ps, pathToWatch, "", nil, nil, &Config{
+		SyncPeriod: 10 * time.Millisecond,
+	}, DefaultWatchConig, nil, fakeHandler, "", nil)
+	assert.NoError(t, err)
+
+	// Create a directory, still nothing is watched.
+	err = os.MkdirAll(logDir1, 0750)
+	assert.NoError(t, err)
+	_, err = os.Create(logfile1)
+	assert.NoError(t, err)
+
+	// There should be only one WatchStart event in the channel so far.
+	ftEvent := <-fakeHandler
+	require.Equal(t, fileTargetEventWatchStart, ftEvent.eventType)
+
+	requireEventually(t, func() bool {
+		return target.getReadersLen() == 1
+	}, "expected 1 tailer to be created")
+
+	require.NoError(t, testutil.GatherAndCompare(registry, bytes.NewBufferString(`
+		# HELP promtail_files_active_total Number of active files.
+		# TYPE promtail_files_active_total gauge
+		promtail_files_active_total 1
+	`), "promtail_files_active_total"))
+
+	// Create two directories - one more than the buffer of fakeHandler,
+	// so that the file target hands until we call Stop().
+	err = os.MkdirAll(logDir2, 0750)
+	assert.NoError(t, err)
+	_, err = os.Create(logfile2)
+	assert.NoError(t, err)
+
+	err = os.MkdirAll(logDir3, 0750)
+	assert.NoError(t, err)
+	_, err = os.Create(logfile3)
+	assert.NoError(t, err)
+
+	// Wait until the file target is waiting on a channel send due to a full channel buffer.
+	requireEventually(t, func() bool {
+		return len(fakeHandler) == 1
+	}, "expected an event in the fakeHandler channel")
+
+	// If FileHandler works well, then it will stop waiting for
+	// the blocked fakeHandler and stop cleanly.
+	// This is why this time we don't drain fakeHandler.
+	target.Stop()
+	ps.Stop()
```

Comment:
This test still isn't 100% correct IMO.

If you copy/paste it and execute it against the current main Loki branch, it will never succeed or fail except due to the general 10 minute timelimit on `go test` runs, because of the same original issue that `Stop` calls have a deadlock currently. 

It should fail within a few seconds against `main`, and then pass almost instantaneously with the change from your original commit.

Reply:
Just to clarify - you are stating that the test is not correct because it fails due to a timeout, and not because it fails due to a condition somewhere in the test? So you'd like me to make sure we fail the test if Stop() hands for more than a few seconds? I think that's a reasonable ask, but I just want to make sure this is what you'd like me to do.

Reply:
> Just to clarify - you are stating that the test is not correct because it fails due to a timeout, and not because it fails due to a condition somewhere in the test? So you'd like me to make sure we fail the test if Stop() hangs for more than a few seconds? I think that's a reasonable ask, but I just want to make sure this is what you'd like me to do.

More or less, yes. 

Right now against `main` the test fails because in essence the test never actually finishes and reaches an actual success or failure condition. It just reaches the default timeout for `go test` which is 10 minutes. 

The goal of your change is to make `Stop` and related functions not block deadlock via allowing select on the closing of the related channel. By wrapping these two `Stop()` calls in a go routine and then changing the last `require.NoError` to a wrapper with `require.Eventually` the test should fail due to `Stop` not returning in a timely manner, and then with your change the test would pass.

Reply:
I updated the test to wrap `Stop()` with `requireEventually`. I don't think there is a need to run `Stop()` in a goroutine.

Reply:
> I updated the test to wrap Stop() with requireEventually. I don't think there is a need to run Stop() in a goroutine.

your solution is simpler :+1: